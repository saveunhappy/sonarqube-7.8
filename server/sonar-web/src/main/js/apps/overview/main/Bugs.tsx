/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import enhance, { ComposedProps } from './enhance';
import ApplicationLeakPeriodLegend from '../components/ApplicationLeakPeriodLegend';
import BugIcon from '../../../components/icons-components/BugIcon';
import DocTooltip from '../../../components/docs/DocTooltip';
import DateFromNow from '../../../components/intl/DateFromNow';
import LeakPeriodLegend from '../components/LeakPeriodLegend';
import { getMetricName } from '../utils';
import { isLongLivingBranch } from '../../../helpers/branches';
import { translateWithParameters } from '../../../helpers/l10n';

export class Bugs extends React.PureComponent<ComposedProps> {
  renderHeader() {
    return this.props.renderHeader('Reliability');
  }

  renderTimelineStartDate(historyStartDate?: Date) {
    if (!historyStartDate) {
      return undefined;
    }
    return (
      <DateFromNow date={historyStartDate}>
        {fromNow => (
          <span className="overview-domain-timeline-date">
            {translateWithParameters('overview.started_x', fromNow)}
          </span>
        )}
      </DateFromNow>
    );
  }

  renderTimeline(range: string, historyStartDate?: Date) {
    return this.props.renderTimeline('bugs', range, this.renderTimelineStartDate(historyStartDate));
  }

  renderLeak() {
    const { branchLike, component, leakPeriod } = this.props;
    if (!leakPeriod) {
      return null;
    }

    return (
      <div className="overview-domain-leak">
        {component.qualifier === 'APP' ? (
          <ApplicationLeakPeriodLegend
            branch={isLongLivingBranch(branchLike) ? branchLike : undefined}
            component={component}
          />
        ) : (
          <LeakPeriodLegend period={leakPeriod} />
        )}

        <div className="overview-domain-measures">
          <div className="overview-domain-measure">
            <div className="overview-domain-measure-value">
              <span className="offset-left">{this.props.renderIssues('new_bugs', 'BUG')}</span>
              {this.props.renderRating('new_reliability_rating')}
            </div>
            <div className="overview-domain-measure-label">
              <BugIcon className="little-spacer-right" />
              {getMetricName('new_bugs')}
            </div>
          </div>
        </div>
        {this.renderTimeline('after')}
      </div>
    );
  }

  renderNutshell() {
    return (
      <div className="overview-domain-nutshell">
        <div className="overview-domain-measures">
          <div className="overview-domain-measure">
            <div className="overview-domain-measure-value">
              <span className="offset-left">{this.props.renderIssues('bugs', 'BUG')}</span>
              {this.props.renderRating('reliability_rating')}
            </div>
            <div className="overview-domain-measure-label display-flex-center display-flex-justify-center">
              <BugIcon className="little-spacer-right " />
              {getMetricName('bugs')}
              <DocTooltip
                className="little-spacer-left"
                doc={import(/* webpackMode: "eager" */ 'Docs/tooltips/metrics/bugs.md')}
              />
            </div>
            {this.props.renderHistoryLink('bugs')}
          </div>
        </div>
        {this.renderTimeline('before', this.props.historyStartDate)}
      </div>
    );
  }

  render() {
    const { measures } = this.props;
    const bugsMeasure = measures.find(measure => measure.metric.key === 'bugs');
    if (!bugsMeasure) {
      return null;
    }
    return (
      <div className="overview-card">
        {this.renderHeader()}

        <div className="overview-domain-panel">
          {this.renderNutshell()}
          {this.renderLeak()}
        </div>
      </div>
    );
  }
}

export default enhance(Bugs);
