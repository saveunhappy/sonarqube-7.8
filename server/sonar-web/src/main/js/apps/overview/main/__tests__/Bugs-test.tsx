/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import Bugs from '../Bugs';
import { ComposedProps } from '../enhance';
import {
  mockComponent,
  mockMainBranch,
  mockMeasureEnhanced,
  mockMetric
} from '../../../../helpers/testMocks';

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot();
});

function shallowRender(props: Partial<ComposedProps> = {}) {
  return shallow(
    <Bugs
      branchLike={mockMainBranch()}
      component={mockComponent()}
      history={{ bugs: [] }}
      historyStartDate={new Date('2019-01-14T15:44:51.000Z')}
      leakPeriod={{ index: 1, mode: 'days' } as T.Period}
      measures={[
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'bugs',
            key: 'bugs',
            name: 'Bugs',
            type: 'INT'
          }),
          value: '5'
        }),
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'new_bugs',
            key: 'new_bugs',
            name: 'New Bugs',
            type: 'INT'
          }),
          value: '2'
        }),
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'reliability_rating',
            key: 'reliability_rating',
            name: 'Reliability',
            type: 'RATING'
          }),
          value: '1.0'
        }),
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'new_reliability_rating',
            key: 'new_reliability_rating',
            name: 'New Reliability',
            type: 'RATING'
          }),
          value: '2.0'
        })
      ]}
      {...props}
    />
  ).dive();
}
