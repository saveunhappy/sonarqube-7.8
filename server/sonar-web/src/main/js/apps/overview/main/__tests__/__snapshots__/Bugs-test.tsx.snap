// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="overview-card"
>
  <div
    className="overview-card-header"
  >
    <div
      className="overview-title"
    >
      <span>
        metric_domain.Reliability
      </span>
      <Link
        className="spacer-left small"
        onlyActiveOnIndex={false}
        style={Object {}}
        to={
          Object {
            "pathname": "/component_measures",
            "query": Object {
              "id": "my-project",
              "metric": "Reliability",
            },
          }
        }
      >
        layout.measures
      </Link>
    </div>
  </div>
  <div
    className="overview-domain-panel"
  >
    <div
      className="overview-domain-nutshell"
    >
      <div
        className="overview-domain-measures"
      >
        <div
          className="overview-domain-measure"
        >
          <div
            className="overview-domain-measure-value"
          >
            <span
              className="offset-left"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to={
                  Object {
                    "pathname": "/project/issues",
                    "query": Object {
                      "id": "my-project",
                      "resolved": "false",
                      "types": "BUG",
                    },
                  }
                }
              >
                5
              </Link>
            </span>
            <Tooltip
              overlay="metric.reliability_rating.tooltip.A"
            >
              <div
                className="overview-domain-measure-sup"
              >
                <DrilldownLink
                  branchLike={
                    Object {
                      "analysisDate": "2018-01-01",
                      "isMain": true,
                      "name": "master",
                    }
                  }
                  className="link-no-underline"
                  component="my-project"
                  metric="reliability_rating"
                >
                  <Rating
                    value="1.0"
                  />
                </DrilldownLink>
              </div>
            </Tooltip>
          </div>
          <div
            className="overview-domain-measure-label display-flex-center display-flex-justify-center"
          >
            <BugIcon
              className="little-spacer-right "
            />
            overview.metric.bugs
            <DocTooltip
              className="little-spacer-left"
              doc={Promise {}}
            />
          </div>
          <Link
            className="overview-domain-measure-history-link"
            onlyActiveOnIndex={false}
            style={Object {}}
            to={
              Object {
                "pathname": "/project/activity",
                "query": Object {
                  "custom_metrics": "bugs",
                  "graph": "custom",
                  "id": "my-project",
                },
              }
            }
          >
            <HistoryIcon />
            <span>
              project_activity.page
            </span>
          </Link>
        </div>
      </div>
      <div
        className="overview-domain-timeline"
      >
        <Timeline
          history={Array []}
        />
        <DateFromNow
          date={2019-01-14T15:44:51.000Z}
        >
          <Component />
        </DateFromNow>
      </div>
    </div>
    <div
      className="overview-domain-leak"
    >
      <InjectIntl(LeakPeriodLegend)
        period={
          Object {
            "index": 1,
            "mode": "days",
          }
        }
      />
      <div
        className="overview-domain-measures"
      >
        <div
          className="overview-domain-measure"
        >
          <div
            className="overview-domain-measure-value"
          >
            <span
              className="offset-left"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to={
                  Object {
                    "pathname": "/project/issues",
                    "query": Object {
                      "id": "my-project",
                      "resolved": "false",
                      "sinceLeakPeriod": "true",
                      "types": "BUG",
                    },
                  }
                }
              >
                1
              </Link>
            </span>
            <Tooltip
              overlay="metric.reliability_rating.tooltip.A"
            >
              <div
                className="overview-domain-measure-sup"
              >
                <DrilldownLink
                  branchLike={
                    Object {
                      "analysisDate": "2018-01-01",
                      "isMain": true,
                      "name": "master",
                    }
                  }
                  className="link-no-underline"
                  component="my-project"
                  metric="new_reliability_rating"
                >
                  <Rating
                    value="1.0"
                  />
                </DrilldownLink>
              </div>
            </Tooltip>
          </div>
          <div
            className="overview-domain-measure-label"
          >
            <BugIcon
              className="little-spacer-right"
            />
            overview.metric.new_bugs
          </div>
        </div>
      </div>
      <div
        className="overview-domain-timeline"
      >
        <Timeline
          history={Array []}
        />
      </div>
    </div>
  </div>
</div>
`;
